### PR TITLE
Added notebook for PyTrees

### PR DESCRIPTION
**Pull request by gnecula**
Added JAX pytrees notebook
* Also added docstrings to the tree_util module

This PR was already discussed as https://github.com/google/jax/pull/1574, but somehow got lost. I do not quite understand how, I thought I have merged it.

### CODE REVIEW

**Comment by mattjj** [comment]
Looking at the diff on #1574, it looks like the notebook just wasn't included, so I guess it was never merged into master. Maybe lost in a rebase somehow? No idea :)

File: jax/tree_util.py [hunk 1]
```diff
@@ -44,9 +47,70 @@
...
+  """Reconstructs a pytree from the treedef and the leaves.
+
+  The inverse of `tree_flatten`.
+
+  Args:
+    treedef: the treedef to reconstruct
+    leaves: the list of leaves to use for reconstruction. The list must
+      match the leaves of the treedef.
+  Returns:
+    The reconstructed pytree, containing the `leaves` places in the
```
Comment (mattjj) [inline jax/tree_util.py]: typo: "places" -> "placed"

**Review by mattjj — APPROVED** [review]
LGTM, thanks!

File: jax/tree_util.py (hunk 1 above)
Comment (gnecula) [reply jax/tree_util.py]: Done

**Comment by gnecula** [comment]
Fixes #1552

**Pull request merged**